src/transformStateWithClones.js

```diff
@@ -7,7 +7,30 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        break;
+      case 'removeProperties':
+        const newState = { ...currentState };
+
+        for (const key of action.keysToRemove) {
+          delete newState[key];
+        }
+        currentState = newState;
+        break;
+      case 'clear':
+        currentState = {};
+        break;
+    }
```

Comment:
it's good practice to have `default` case in a switch as is said in the previous comment

If you don't know what to do in default case you can throw an error - it's common way to handle it

for example:

```suggestion
    }
    
    default:
      throw new Error(`Unknown action type: ${action.type}`);
```